src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs

```diff
@@ -1786,64 +1790,103 @@ public static Matrix4x4 Transpose(Matrix4x4 matrix)
-            result.M33 = matrix1.M33 + (matrix2.M33 - matrix1.M33) * amount;
-            result.M34 = matrix1.M34 + (matrix2.M34 - matrix1.M34) * amount;
+                m1Row = Sse.LoadVector128(&matrix1.M31);
+                m2Row = Sse.LoadVector128(&matrix2.M31);
+                Sse.Store(&matrix1.M31, Sse.Add(m1Row, Sse.Multiply(Sse.Subtract(m2Row, m1Row), amountVec)));
 
-            // Fourth row
-            result.M41 = matrix1.M41 + (matrix2.M41 - matrix1.M41) * amount;
-            result.M42 = matrix1.M42 + (matrix2.M42 - matrix1.M42) * amount;
-            result.M43 = matrix1.M43 + (matrix2.M43 - matrix1.M43) * amount;
-            result.M44 = matrix1.M44 + (matrix2.M44 - matrix1.M44) * amount;
-
-            return result;
+                m1Row = Sse.LoadVector128(&matrix1.M41);
+                m2Row = Sse.LoadVector128(&matrix2.M41);
+                Sse.Store(&matrix1.M41, Sse.Add(m1Row, Sse.Multiply(Sse.Subtract(m2Row, m1Row), amountVec)));
+                
+                return matrix1;
+            }
+            else
+            {
+                Matrix4x4 result;
+
+                // First row
+                result.M11 = matrix1.M11 + (matrix2.M11 - matrix1.M11) * amount;
+                result.M12 = matrix1.M12 + (matrix2.M12 - matrix1.M12) * amount;
+                result.M13 = matrix1.M13 + (matrix2.M13 - matrix1.M13) * amount;
+                result.M14 = matrix1.M14 + (matrix2.M14 - matrix1.M14) * amount;
+
+                // Second row
+                result.M21 = matrix1.M21 + (matrix2.M21 - matrix1.M21) * amount;
+                result.M22 = matrix1.M22 + (matrix2.M22 - matrix1.M22) * amount;
+                result.M23 = matrix1.M23 + (matrix2.M23 - matrix1.M23) * amount;
+                result.M24 = matrix1.M24 + (matrix2.M24 - matrix1.M24) * amount;
+
+                // Third row
+                result.M31 = matrix1.M31 + (matrix2.M31 - matrix1.M31) * amount;
+                result.M32 = matrix1.M32 + (matrix2.M32 - matrix1.M32) * amount;
+                result.M33 = matrix1.M33 + (matrix2.M33 - matrix1.M33) * amount;
+                result.M34 = matrix1.M34 + (matrix2.M34 - matrix1.M34) * amount;
+
+                // Fourth row
+                result.M41 = matrix1.M41 + (matrix2.M41 - matrix1.M41) * amount;
+                result.M42 = matrix1.M42 + (matrix2.M42 - matrix1.M42) * amount;
+                result.M43 = matrix1.M43 + (matrix2.M43 - matrix1.M43) * amount;
+                result.M44 = matrix1.M44 + (matrix2.M44 - matrix1.M44) * amount;
+
+                return result;
+            }
         }
 
         /// <summary>
         /// Returns a new matrix with the negated elements of the given matrix.
         /// </summary>
         /// <param name="value">The source matrix.</param>
         /// <returns>The negated matrix.</returns>
-        public static Matrix4x4 Negate(Matrix4x4 value)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
I'm not sure about this. The code is probably small enough to be inlined in the hardware accelerated case, but not necessarily in the software case.

Reply:
Same goes for other new `AggressiveInlining` attributes

Reply:
What is the codegen you are seeing here?

Reply:
@tannergooding Negate is not inlined 🙁 https://gist.github.com/EgorBo/a6cc52d5523d7a5fcf6fb1adfa5dfbc0 
`Negate_new` is a benchmark method that does `return Matrix4x4X.Negate(m1x);`

Reply:
thus current numbers are:

|        Method |      Mean |     Error |    StdDev | Scaled |
|-------------- |----------:|----------:|----------:|-------:|
|    Negate_old | 13.385 ns | 0.0781 ns | 0.0693 ns |   1.00 |
|    Negate_new |  3.364 ns | 0.0230 ns | 0.0192 ns |   0.25 |
| Negate_new_op |  3.023 ns | 0.0257 ns | 0.0240 ns |   0.23 |

Reply:
I'm seeing that for `var negated = Matrix4x4.Negate(Matrix4x4.Identity);` where `public static Matrix4x4 Negate(Matrix4x4 value) => -value`, there is only a single call and it contains the relevant SSE code (there isn't a call, with an immediate nested call).

I'm building CoreCLR locally to get the jit-utils diff and will post sometime after lunch.

Reply:
um.. I expected `Negate_new()` to just call op_UnaryNegation directly  
![image](https://user-images.githubusercontent.com/523221/44167843-f2663f00-a0d7-11e8-9165-8f2c439dcaa1.png)

Reply:
Ok, using `jitutils` to get PMI diffs, and for `public static unsafe Matrix4x4 Negate(Matrix4x4 value) => -value;` and `public static unsafe Matrix4x4 operator -(Matrix4x4 value)` containing the actual implementation....

The following two tests (aside from not being "correct"):
```C#
[Fact]
public void Matrix4x4NegateTest()
{
    Matrix4x4 m = GenerateMatrixNumberFrom1To16();
    var actual = Matrix4x4.Negate(m);
    Assert.Equal(m, actual);
}

[Fact]
public void Matrix4x4op_NegateTest()
{
    Matrix4x4 m = GenerateMatrixNumberFrom1To16();
    var actual = -m;
    Assert.Equal(m, actual);
}
```

Both generate a call directly to `call     Matrix4x4:op_UnaryNegation(struct):struct`:
```asm
; Assembly listing for method Matrix4x4Tests:Matrix4x4NegateTest():this
; Emitting BLENDED_CODE for X64 CPU with AVX
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;* V00 this         [V00    ] (  0,  0   )     ref  ->  zero-ref    this class-hnd
;  V01 loc0         [V01    ] (  3,  3   )  struct (64) [rsp+0x1A8]   do-not-enreg[XSB] must-init addr-exposed
;  V02 loc1         [V02    ] (  2,  2   )  struct (64) [rsp+0x168]   do-not-enreg[XSB] must-init addr-exposed
;  V03 OutArgs      [V03    ] (  1,  1   )  lclBlk (32) [rsp+0x00]  
;  V04 tmp1         [V04,T01] (  2,  4   )  struct (64) [rsp+0x128]   do-not-enreg[SB]
;  V05 tmp2         [V05,T02] (  2,  4   )  struct (64) [rsp+0xE8]   do-not-enreg[SB]
;  V06 tmp3         [V06,T03] (  2,  4   )  struct (64) [rsp+0xA8]   do-not-enreg[SB]
;  V07 tmp4         [V07,T00] (  3,  6   )     ref  ->  rsi         class-hnd exact
;  V08 tmp5         [V08    ] (  4,  8   )  struct (64) [rsp+0x68]   do-not-enreg[XSB] addr-exposed
;  V09 tmp6         [V09,T04] (  2,  4   )    long  ->  rcx        
;  V10 tmp7         [V10    ] (  2,  4   )  struct (64) [rsp+0x28]   do-not-enreg[XSB] addr-exposed
;
; Lcl frame size = 488
G_M53037_IG01:
       push     rdi
       push     rsi
       sub      rsp, 488
       vzeroupper 
       mov      rsi, rcx
       lea      rdi, [rsp+168H]
       mov      ecx, 32
       xor      rax, rax
       rep stosd 
       mov      rcx, rsi
G_M53037_IG02:
       lea      rcx, bword ptr [rsp+1A8H]
       call     Matrix4x4Tests:GenerateMatrixNumberFrom1To16():struct
       vmovdqu  xmm0, qword ptr [rsp+1A8H]
       vmovdqu  qword ptr [rsp+128H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+1B8H]
       vmovdqu  qword ptr [rsp+138H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+1C8H]
       vmovdqu  qword ptr [rsp+148H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+1D8H]
       vmovdqu  qword ptr [rsp+158H], xmm0
       lea      rcx, bword ptr [rsp+168H]
       vmovdqu  xmm0, qword ptr [rsp+128H]
       vmovdqu  qword ptr [rsp+68H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+138H]
       vmovdqu  qword ptr [rsp+78H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+148H]
       vmovdqu  qword ptr [rsp+88H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+158H]
       vmovdqu  qword ptr [rsp+98H], xmm0
       lea      rdx, bword ptr [rsp+68H]
       call     Matrix4x4:op_UnaryNegation(struct):struct
       vmovdqu  xmm0, qword ptr [rsp+1A8H]
       vmovdqu  qword ptr [rsp+E8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+1B8H]
       vmovdqu  qword ptr [rsp+F8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+1C8H]
       vmovdqu  qword ptr [rsp+108H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+1D8H]
       vmovdqu  qword ptr [rsp+118H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+168H]
       vmovdqu  qword ptr [rsp+A8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+178H]
       vmovdqu  qword ptr [rsp+B8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+188H]
       vmovdqu  qword ptr [rsp+C8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+198H]
       vmovdqu  qword ptr [rsp+D8H], xmm0
       mov      rcx, 0xD1FFAB1E
       call     CORINFO_HELP_NEWSFAST
       mov      rsi, rax
       mov      rcx, rsi
       xor      rdx, rdx
       call     AssertEqualityComparer`1:.ctor(ref):this
       vmovdqu  xmm0, qword ptr [rsp+E8H]
       vmovdqu  qword ptr [rsp+68H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+F8H]
       vmovdqu  qword ptr [rsp+78H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+108H]
       vmovdqu  qword ptr [rsp+88H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+118H]
       vmovdqu  qword ptr [rsp+98H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+A8H]
       vmovdqu  qword ptr [rsp+28H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+B8H]
       vmovdqu  qword ptr [rsp+38H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+C8H]
       vmovdqu  qword ptr [rsp+48H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+D8H]
       vmovdqu  qword ptr [rsp+58H], xmm0
G_M53037_IG03:
       lea      rcx, bword ptr [rsp+68H]
       lea      rdx, bword ptr [rsp+28H]
       mov      r8, rsi
       call     Assert:Equal(struct,struct,ref)
       nop      
G_M53037_IG04:
       add      rsp, 488
       pop      rsi
       pop      rdi
       ret      
; Total bytes of code 579, prolog size 35 for method Matrix4x4Tests:Matrix4x4NegateTest():this
; ============================================================
Unwind Info:
  >> Start offset   : 0x000000 (not in unwind data)
  >>   End offset   : 0xd1ffab1e (not in unwind data)
  Version           : 1
  Flags             : 0x00
  SizeOfProlog      : 0x09
  CountOfUnwindCodes: 4
  FrameRegister     : none (0)
  FrameOffset       : N/A (no FrameRegister) (Value=0)
  UnwindCodes       :
    CodeOffset: 0x09 UnwindOp: UWOP_ALLOC_LARGE (1)     OpInfo: 0 - Scaled small  
      Size: 61 * 8 = 488 = 0x001E8
    CodeOffset: 0x02 UnwindOp: UWOP_PUSH_NONVOL (0)     OpInfo: rsi (6)
    CodeOffset: 0x01 UnwindOp: UWOP_PUSH_NONVOL (0)     OpInfo: rdi (7)
```
and
```asm
; Assembly listing for method Matrix4x4Tests:Matrix4x4op_NegateTest():this
; Emitting BLENDED_CODE for X64 CPU with AVX
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;* V00 this         [V00    ] (  0,  0   )     ref  ->  zero-ref    this class-hnd
;  V01 loc0         [V01    ] (  3,  3   )  struct (64) [rsp+0x168]   do-not-enreg[XSB] must-init addr-exposed
;  V02 loc1         [V02    ] (  2,  2   )  struct (64) [rsp+0x128]   do-not-enreg[XSB] must-init addr-exposed
;  V03 OutArgs      [V03    ] (  1,  1   )  lclBlk (32) [rsp+0x00]  
;  V04 tmp1         [V04,T01] (  2,  4   )  struct (64) [rsp+0xE8]   do-not-enreg[SB]
;  V05 tmp2         [V05,T02] (  2,  4   )  struct (64) [rsp+0xA8]   do-not-enreg[SB]
;  V06 tmp3         [V06,T00] (  3,  6   )     ref  ->  rsi         class-hnd exact
;  V07 tmp4         [V07    ] (  4,  8   )  struct (64) [rsp+0x68]   do-not-enreg[XSB] addr-exposed
;  V08 tmp5         [V08,T03] (  2,  4   )    long  ->  rcx        
;  V09 tmp6         [V09    ] (  2,  4   )  struct (64) [rsp+0x28]   do-not-enreg[XSB] addr-exposed
;
; Lcl frame size = 424
G_M27564_IG01:
       push     rdi
       push     rsi
       sub      rsp, 424
       vzeroupper 
       mov      rsi, rcx
       lea      rdi, [rsp+128H]
       mov      ecx, 32
       xor      rax, rax
       rep stosd 
       mov      rcx, rsi
G_M27564_IG02:
       lea      rcx, bword ptr [rsp+168H]
       call     Matrix4x4Tests:GenerateMatrixNumberFrom1To16():struct
       lea      rcx, bword ptr [rsp+128H]
       vmovdqu  xmm0, qword ptr [rsp+168H]
       vmovdqu  qword ptr [rsp+68H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+178H]
       vmovdqu  qword ptr [rsp+78H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+188H]
       vmovdqu  qword ptr [rsp+88H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+198H]
       vmovdqu  qword ptr [rsp+98H], xmm0
       lea      rdx, bword ptr [rsp+68H]
       call     Matrix4x4:op_UnaryNegation(struct):struct
       vmovdqu  xmm0, qword ptr [rsp+168H]
       vmovdqu  qword ptr [rsp+E8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+178H]
       vmovdqu  qword ptr [rsp+F8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+188H]
       vmovdqu  qword ptr [rsp+108H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+198H]
       vmovdqu  qword ptr [rsp+118H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+128H]
       vmovdqu  qword ptr [rsp+A8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+138H]
       vmovdqu  qword ptr [rsp+B8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+148H]
       vmovdqu  qword ptr [rsp+C8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+158H]
       vmovdqu  qword ptr [rsp+D8H], xmm0
       mov      rcx, 0xD1FFAB1E
       call     CORINFO_HELP_NEWSFAST
       mov      rsi, rax
       mov      rcx, rsi
       xor      rdx, rdx
       call     AssertEqualityComparer`1:.ctor(ref):this
       vmovdqu  xmm0, qword ptr [rsp+E8H]
       vmovdqu  qword ptr [rsp+68H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+F8H]
       vmovdqu  qword ptr [rsp+78H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+108H]
       vmovdqu  qword ptr [rsp+88H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+118H]
       vmovdqu  qword ptr [rsp+98H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+A8H]
       vmovdqu  qword ptr [rsp+28H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+B8H]
       vmovdqu  qword ptr [rsp+38H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+C8H]
       vmovdqu  qword ptr [rsp+48H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+D8H]
       vmovdqu  qword ptr [rsp+58H], xmm0
       lea      rcx, bword ptr [rsp+68H]
       lea      rdx, bword ptr [rsp+28H]
       mov      r8, rsi
       call     Assert:Equal(struct,struct,ref)
       nop      
G_M27564_IG03:
       add      rsp, 424
       pop      rsi
       pop      rdi
       ret      
; Total bytes of code 499, prolog size 35 for method Matrix4x4Tests:Matrix4x4op_NegateTest():this
; ============================================================
Unwind Info:
  >> Start offset   : 0x000000 (not in unwind data)
  >>   End offset   : 0xd1ffab1e (not in unwind data)
  Version           : 1
  Flags             : 0x00
  SizeOfProlog      : 0x09
  CountOfUnwindCodes: 4
  FrameRegister     : none (0)
  FrameOffset       : N/A (no FrameRegister) (Value=0)
  UnwindCodes       :
    CodeOffset: 0x09 UnwindOp: UWOP_ALLOC_LARGE (1)     OpInfo: 0 - Scaled small  
      Size: 53 * 8 = 424 = 0x001A8
    CodeOffset: 0x02 UnwindOp: UWOP_PUSH_NONVOL (0)     OpInfo: rsi (6)
    CodeOffset: 0x01 UnwindOp: UWOP_PUSH_NONVOL (0)     OpInfo: rdi (7)
```

Reply:
Which is exactly what I expected.

@EgorBo, could you fix them up accordingly, since the JIT already does the right thing.

Reply:
@tannergooding um.. so how should I fix it and do not regress the software fallback (will it be inlined in `.Negate` if SSE is not available?). Is not the current implementation what you meant https://github.com/dotnet/corefx/blob/21ca6ac66e79ee99b3933660fe83ad1f2e07e545/src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs#L1855-L1881

Reply:
I'm not sure what you mean, in both cases `public static unsafe Matrix4x4 Negate(Matrix4x4 value) => -value;` will result in a call directly to `op_Negate`.

The JIT will then determine if `op_Negate` is appropriate to inline or not (currently it does not). It will result in code that is simpler to maintain and produces the same codegen as is being produced today (for Release builds)